// @spences10/pi-context/dist/store/chunking-helpers.js
import { redact_text } from '@spences10/pi-redact';
import { createHash, randomUUID } from 'node:crypto';
import { chunk_text, count_lines, make_preview, should_index_text, summarize_source, } from '../text.js';
import { scoped_filter, } from './query-helpers.js';
function find_duplicate_source(db, state, content_hash, scope) {
    const scoped = scoped_filter(state, 'context_sources', scope);
    const filters = [
        'context_sources.content_hash = ?',
        ...scoped.where,
    ];
    const params = [
        content_hash,
        ...scoped.params,
    ];
    const row = db
        .prepare(`
			SELECT
				context_sources.id,
				COUNT(context_chunks.id) as chunk_count,
				(
					SELECT first_chunk.id FROM context_chunks first_chunk
					WHERE first_chunk.source_id = context_sources.id
					ORDER BY first_chunk.ordinal LIMIT 1
				) as first_chunk_id
			FROM context_sources
			LEFT JOIN context_chunks ON context_chunks.source_id = context_sources.id
			WHERE ${filters.join(' AND ')}
			GROUP BY context_sources.id
			ORDER BY context_sources.created_at DESC
			LIMIT 1
		`)
        .get(...params);
    return row ?? null;
}
export function store_context_input(db, state, limits, input) {
    const redaction = redact_text(input.text);
    const text = redaction.redacted;
    if (!input.force && !should_index_text(text, limits))
        return null;
    const bytes = Buffer.byteLength(text, 'utf8');
    const lines = count_lines(text);
    const created_at = Date.now();
    const content_hash = createHash('sha256')
        .update(text)
        .digest('hex');
    const session_id = input.session_id ?? state.session_id;
    const project_path = input.project_path ?? state.project_path;
    const preview = make_preview(text);
    const duplicate = find_duplicate_source(db, state, content_hash, {
        project_path,
    });
    if (duplicate) {
        const provisional = {
            source_id: duplicate.id,
            bytes,
            lines,
            preview,
            receipt: '',
            chunk_count: duplicate.chunk_count,
            first_chunk_id: duplicate.first_chunk_id,
            returned_bytes: 0,
            project_path,
            session_id,
            deduped: true,
        };
        const receipt = summarize_source(provisional, input.tool_name);
        const returned_bytes = Buffer.byteLength(receipt, 'utf8');
        db.prepare('UPDATE context_sources SET returned_byte_count = returned_byte_count + ? WHERE id = ?').run(returned_bytes, duplicate.id);
        return { ...provisional, receipt, returned_bytes };
    }
    const source_id = `ctx_${Date.now().toString(36)}_${randomUUID().slice(0, 8)}`;
    const chunks = chunk_text(text, source_id);
    const preview_bytes = Buffer.byteLength(preview, 'utf8');
    const insert = db.prepare(`
		INSERT INTO context_sources (
			id, session_id, project_path, tool_name, input_summary, created_at,
			byte_count, line_count, content_hash, preview_byte_count, returned_byte_count
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0)
	`);
    const insert_chunk = db.prepare(`
		INSERT INTO context_chunks (id, source_id, ordinal, title, content, byte_count)
		VALUES (?, ?, ?, ?, ?, ?)
	`);
    const update_returned = db.prepare('UPDATE context_sources SET returned_byte_count = ? WHERE id = ?');
    db.exec('BEGIN');
    try {
        insert.run(source_id, session_id, project_path, input.tool_name, input.input_summary ?? null, created_at, bytes, lines, content_hash, preview_bytes);
        for (const chunk of chunks) {
            insert_chunk.run(chunk.id, chunk.source_id, chunk.ordinal, chunk.title, chunk.content, chunk.byte_count);
        }
        const provisional = {
            source_id,
            bytes,
            lines,
            preview,
            receipt: '',
            chunk_count: chunks.length,
            first_chunk_id: chunks[0]?.id ?? null,
            returned_bytes: 0,
            project_path,
            session_id,
        };
        const receipt = summarize_source(provisional, input.tool_name);
        const returned_bytes = Buffer.byteLength(receipt, 'utf8');
        update_returned.run(returned_bytes, source_id);
        db.exec('COMMIT');
        return { ...provisional, receipt, returned_bytes };
    }
    catch (error) {
        db.exec('ROLLBACK');
        throw error;
    }
}
