// @spences10/pi-context/dist/store/purge-helpers.js
import { parse_context_retention_policy } from '../policy.js';
export function purge_to_max_stored_bytes(db, max_bytes) {
    const total_row = db
        .prepare('SELECT COALESCE(SUM(byte_count), 0) as bytes FROM context_sources')
        .get();
    let total = total_row.bytes;
    if (total <= max_bytes)
        return 0;
    const rows = db
        .prepare('SELECT id, byte_count FROM context_sources ORDER BY created_at ASC')
        .all();
    const delete_source = db.prepare('DELETE FROM context_sources WHERE id = ?');
    let deleted = 0;
    for (const row of rows) {
        if (total <= max_bytes)
            break;
        const result = delete_source.run(row.id);
        if (Number(result.changes ?? 0) > 0) {
            deleted += 1;
            total -= row.byte_count;
        }
    }
    return deleted;
}
export function purge_context_sources(db, options = {}) {
    const filters = [];
    const params = [];
    if (options.source_id) {
        filters.push('id = ?');
        params.push(options.source_id);
    }
    if (options.project_path === null) {
        filters.push('project_path IS NULL');
    }
    else if (options.project_path !== undefined) {
        filters.push('project_path = ?');
        params.push(options.project_path);
    }
    if (options.session_id === null) {
        filters.push('session_id IS NULL');
    }
    else if (options.session_id !== undefined) {
        filters.push('session_id = ?');
        params.push(options.session_id);
    }
    const days = options.older_than_days;
    if (days !== undefined) {
        filters.push('created_at < ?');
        params.push(Date.now() - days * 24 * 60 * 60 * 1000);
    }
    if (filters.length === 0)
        return { deleted: 0 };
    const result = db
        .prepare(`DELETE FROM context_sources WHERE ${filters.join(' AND ')}`)
        .run(...params);
    return {
        deleted: Number(result.changes ?? 0),
        source_id: options.source_id,
        project_path: options.project_path,
        session_id: options.session_id,
        older_than_days: options.older_than_days,
    };
}
export function cleanup_context_sources(db, policy = parse_context_retention_policy()) {
    const age_deleted = policy.retention_days === null
        ? 0
        : purge_context_sources(db, {
            older_than_days: policy.retention_days,
        }).deleted;
    const size_deleted = policy.max_bytes
        ? purge_to_max_stored_bytes(db, policy.max_bytes)
        : 0;
    return {
        deleted: age_deleted + size_deleted,
        age_deleted,
        size_deleted,
        policy,
    };
}
