// @spences10/pi-context/dist/store/query-helpers.js
export function scoped_filter(state, alias, options = {}) {
    const where = [];
    const params = [];
    if (options.session_id === null) {
        where.push(`${alias}.session_id IS NULL`);
    }
    else if (options.session_id !== undefined) {
        where.push(`${alias}.session_id = ?`);
        params.push(options.session_id);
    }
    else if (!options.global && state.session_id) {
        where.push(`${alias}.session_id = ?`);
        params.push(state.session_id);
    }
    if (options.project_path === null) {
        where.push(`${alias}.project_path IS NULL`);
    }
    else if (options.project_path !== undefined) {
        where.push(`${alias}.project_path = ?`);
        params.push(options.project_path);
    }
    else if (!options.global &&
        where.length === 0 &&
        state.project_path) {
        where.push(`${alias}.project_path = ?`);
        params.push(state.project_path);
    }
    return { where, params };
}
export function list_context_sources(db, state, options = {}) {
    const limit = Math.max(1, Math.min(options.limit ?? 10, 50));
    const offset = Math.max(0, options.offset ?? 0);
    const scoped = scoped_filter(state, 'context_sources', options);
    const filters = [...scoped.where];
    const params = [...scoped.params];
    if (options.source_id) {
        filters.push('context_sources.id = ?');
        params.push(options.source_id);
    }
    if (options.tool_name) {
        filters.push('context_sources.tool_name = ?');
        params.push(options.tool_name);
    }
    if (options.newer_than_days !== undefined) {
        filters.push('context_sources.created_at >= ?');
        params.push(Date.now() - options.newer_than_days * 24 * 60 * 60 * 1000);
    }
    if (options.older_than_days !== undefined) {
        filters.push('context_sources.created_at < ?');
        params.push(Date.now() - options.older_than_days * 24 * 60 * 60 * 1000);
    }
    params.push(limit, offset);
    const where_clause = filters.length
        ? `WHERE ${filters.join(' AND ')}`
        : '';
    const stmt = db.prepare(`
		SELECT
			context_sources.id as source_id,
			context_sources.created_at,
			context_sources.project_path,
			context_sources.session_id,
			context_sources.tool_name,
			context_sources.input_summary,
			context_sources.byte_count,
			context_sources.line_count,
			COUNT(context_chunks.id) as chunk_count,
			(
				SELECT title FROM context_chunks first_chunk
				WHERE first_chunk.source_id = context_sources.id
				ORDER BY ordinal LIMIT 1
			) as first_chunk_title,
			(
				SELECT substr(content, 1, 240) FROM context_chunks first_chunk
				WHERE first_chunk.source_id = context_sources.id
				ORDER BY ordinal LIMIT 1
			) as preview
		FROM context_sources
		LEFT JOIN context_chunks ON context_chunks.source_id = context_sources.id
		${where_clause}
		GROUP BY context_sources.id
		ORDER BY context_sources.created_at DESC
		LIMIT ? OFFSET ?
	`);
    return stmt.all(...params).map((row) => ({
        source_id: row.source_id,
        created_at: row.created_at,
        project_path: row.project_path,
        session_id: row.session_id,
        tool_name: row.tool_name,
        input_summary: row.input_summary,
        bytes: row.byte_count,
        lines: row.line_count,
        chunk_count: row.chunk_count,
        first_chunk_title: row.first_chunk_title,
        preview: row.preview,
    }));
}
export function count_context_stats(db, state, options) {
    const scoped = scoped_filter(state, 'context_sources', options);
    const where_clause = scoped.where.length
        ? `WHERE ${scoped.where.join(' AND ')}`
        : '';
    const source = db
        .prepare(`
			SELECT
				COUNT(*) as sources,
				COALESCE(SUM(byte_count), 0) as bytes_stored,
				COALESCE(SUM(returned_byte_count), 0) as bytes_returned,
				MIN(created_at) as oldest_created_at,
				MAX(created_at) as newest_created_at
			FROM context_sources
			${where_clause}
		`)
        .get(...scoped.params);
    const chunks = db
        .prepare(`
			SELECT COUNT(context_chunks.id) as chunks
			FROM context_chunks
			JOIN context_sources ON context_sources.id = context_chunks.source_id
			${where_clause}
		`)
        .get(...scoped.params);
    return { ...source, chunks: chunks.chunks };
}
