// @spences10/pi-context/dist/store/search-helpers.js
import { escape_fts5_query, relaxed_fts5_query } from '../text.js';
import { scoped_filter, } from './query-helpers.js';
function escape_regexp(value) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
export function chunk_reference_to_ordinal(source_id, chunk_id) {
    const trimmed = chunk_id.trim();
    const legacy_match = new RegExp(`^${escape_regexp(source_id)}:chunk:(\\d+)$`).exec(trimmed);
    if (legacy_match) {
        const value = Number.parseInt(legacy_match[1], 10);
        if (!Number.isSafeInteger(value))
            return null;
        return value <= 0 ? 1 : value;
    }
    if (!/^\d+$/.test(trimmed))
        return null;
    const value = Number.parseInt(trimmed, 10);
    return Number.isSafeInteger(value) && value > 0 ? value : null;
}
function search_match(db, state, match, options, limit) {
    const scoped = options.source_id
        ? { where: [], params: [] }
        : scoped_filter(state, 'context_sources', options);
    const filters = [...scoped.where];
    const params = [match, ...scoped.params];
    if (options.source_id) {
        filters.push('context_sources.id = ?');
        params.push(options.source_id);
    }
    if (options.tool_name) {
        filters.push('context_sources.tool_name = ?');
        params.push(options.tool_name);
    }
    params.push(limit);
    const where_filters = filters.length
        ? ` AND ${filters.join(' AND ')}`
        : '';
    const stmt = db.prepare(`
		SELECT
			context_sources.id,
			context_sources.tool_name,
			context_sources.created_at,
			context_sources.byte_count,
			context_sources.line_count,
			context_chunks.id as chunk_id,
			context_chunks.ordinal,
			context_chunks.title,
			context_chunks.content,
			bm25(context_chunks_fts, 5.0, 1.0) as rank
		FROM context_chunks_fts
		JOIN context_chunks ON context_chunks.rowid = context_chunks_fts.rowid
		JOIN context_sources ON context_sources.id = context_chunks.source_id
		WHERE context_chunks_fts MATCH ?${where_filters}
		ORDER BY rank
		LIMIT ?
	`);
    return stmt.all(...params).map((row) => ({
        source_id: row.id,
        chunk_id: row.chunk_id,
        ordinal: row.ordinal,
        title: row.title,
        content: row.content,
        tool_name: row.tool_name,
        created_at: row.created_at,
        bytes: row.byte_count,
        lines: row.line_count,
        rank: row.rank,
    }));
}
export function search_context_chunks(db, state, query, options = {}) {
    const limit = Math.max(1, Math.min(options.limit ?? 5, 25));
    const strict = escape_fts5_query(query);
    const relaxed = relaxed_fts5_query(query);
    const results = search_match(db, state, strict, options, limit);
    if (results.length >= limit || !relaxed || relaxed === strict)
        return results;
    const seen = new Set(results.map((result) => result.chunk_id));
    for (const result of search_match(db, state, relaxed, options, limit)) {
        if (seen.has(result.chunk_id))
            continue;
        results.push(result);
        seen.add(result.chunk_id);
        if (results.length >= limit)
            break;
    }
    return results;
}
export function chunk_summary(db, state, source_id) {
    const scoped = scoped_filter(state, 'context_sources', {
        global: true,
    });
    const filters = ['context_sources.id = ?', ...scoped.where];
    const params = [
        source_id,
        ...scoped.params,
    ];
    const row = db
        .prepare(`
			SELECT
				context_sources.id as source_id,
				COUNT(context_chunks.id) as chunk_count,
				(
					SELECT first_chunk.id FROM context_chunks first_chunk
					WHERE first_chunk.source_id = context_sources.id
					ORDER BY first_chunk.ordinal LIMIT 1
				) as first_chunk_id,
				(
					SELECT last_chunk.id FROM context_chunks last_chunk
					WHERE last_chunk.source_id = context_sources.id
					ORDER BY last_chunk.ordinal DESC LIMIT 1
				) as last_chunk_id,
				MIN(context_chunks.ordinal) as first_ordinal,
				MAX(context_chunks.ordinal) as last_ordinal
			FROM context_sources
			LEFT JOIN context_chunks ON context_chunks.source_id = context_sources.id
			WHERE ${filters.join(' AND ')}
			GROUP BY context_sources.id
		`)
        .get(...params);
    return row ?? null;
}
export function get_context_chunks(db, state, source_id, chunk_id) {
    const scoped = scoped_filter(state, 'context_sources', {
        global: true,
    });
    const filters = ['context_chunks.source_id = ?', ...scoped.where];
    const params = [
        source_id,
        ...scoped.params,
    ];
    if (chunk_id) {
        const ordinal = chunk_reference_to_ordinal(source_id, chunk_id);
        if (ordinal) {
            filters.push('context_chunks.ordinal = ?');
            params.push(ordinal);
        }
        else {
            filters.push('context_chunks.id = ?');
            params.push(chunk_id);
        }
    }
    const stmt = db.prepare(`
		SELECT
			context_chunks.id,
			context_chunks.source_id,
			context_chunks.ordinal,
			context_chunks.title,
			context_chunks.content,
			context_chunks.byte_count
		FROM context_chunks
		JOIN context_sources ON context_sources.id = context_chunks.source_id
		WHERE ${filters.join(' AND ')}
		ORDER BY context_chunks.ordinal
	`);
    return stmt.all(...params);
}
