// @spences10/pi-context/dist/context-scope.js
export function is_text_content(item) {
    return (!!item &&
        typeof item === 'object' &&
        item.type === 'text' &&
        typeof item.text === 'string');
}
export function summarize_tool_input(input) {
    if (!input || typeof input !== 'object')
        return null;
    try {
        const json = JSON.stringify(input);
        return json.length > 500 ? `${json.slice(0, 497)}...` : json;
    }
    catch {
        return null;
    }
}
export function should_skip_tool(tool_name) {
    // Coverage policy:
    // - context_* tools are retrieval/maintenance output; indexing them would
    //   recurse and make the sidecar harder to reason about.
    // - team output is coordination state, not bulky artifact content; keep it in
    //   team/pirecall surfaces rather than duplicating mailbox/task state here.
    // - MCP receipts are produced before generic tool_result hooks; the hook also
    //   ignores existing [context-sidecar] receipts so direct MCP storage is not
    //   indexed a second time.
    return (tool_name === 'context_search' ||
        tool_name === 'context_get' ||
        tool_name === 'context_list' ||
        tool_name === 'context_stats' ||
        tool_name === 'context_purge' ||
        tool_name === 'team');
}
export function session_id_from_context(ctx) {
    const manager = ctx?.sessionManager;
    return (manager?.getSessionFile?.() ?? manager?.getSessionId?.() ?? null);
}
export function scope_from_context(ctx) {
    return {
        project_path: ctx?.cwd ?? process.cwd(),
        session_id: session_id_from_context(ctx),
    };
}
