// @spences10/pi-context/dist/eval.js
import { mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { ContextStore, should_index_text } from './store.js';
const PROJECT = '/tmp/pi-context-eval-project';
const SESSION = '/tmp/pi-context-eval-session.jsonl';
const search_scenarios = [
    {
        name: 'exact-token-baseline',
        source: 'needle-log',
        query: 'TARGET_VALUE',
        expect: 'TARGET_VALUE appears on line 855',
        limit: 5,
        description: 'Basic exact-token lookup should work before and after.',
    },
    {
        name: 'over-specific-human-query',
        source: 'staging-help',
        query: 'stage specific chunk line commit checks amend author sign-off',
        expect: 'Press space to stage this hunk',
        limit: 5,
        description: 'Human/agent query contains useful words plus absent words; strict AND search used to miss.',
    },
    {
        name: 'literal-or-query',
        source: 'skill-script',
        query: 'run_gh_skill_update OR run_gh_skill_install OR case update',
        expect: 'run_gh_skill_update',
        limit: 5,
        description: 'Queries copied from agent reasoning often include OR; search should treat this as alternatives.',
    },
    {
        name: 'punctuation-symbol-query',
        source: 'code-symbols',
        query: 'registerCommand context-stats',
        expect: "registerCommand('context-stats'",
        limit: 5,
        description: 'Code and command names include punctuation that should not make retrieval brittle.',
    },
    {
        name: 'chunk-boundary-query',
        source: 'chunk-boundary',
        query: 'alpha-token omega-token',
        expect: 'alpha-token',
        limit: 5,
        description: 'Useful terms may be split across chunks; relaxed fallback should still recover likely chunks.',
    },
];
const get_scenarios = [
    {
        name: 'get-first-ordinal-alias',
        source: 'needle-log',
        chunk_id: '1',
        expect: 'before noise line 0',
        description: 'Receipt-led chunk retrieval via ordinal alias should stay reliable.',
    },
];
function make_noise_lines(count, prefix) {
    return Array.from({ length: count }, (_, index) => `${prefix} noise line ${index} ${'x'.repeat(48)}`).join('\n');
}
function source_texts() {
    return {
        'needle-log': {
            tool_name: 'bash',
            text: [
                make_noise_lines(80, 'before'),
                'TARGET_VALUE appears on line 855 with surrounding diagnostic text',
                make_noise_lines(80, 'after'),
            ].join('\n'),
        },
        'staging-help': {
            tool_name: 'bash',
            text: [
                'Interactive git UI help',
                'Press space to stage this hunk.',
                'Press enter to open the selected file.',
                'Use a to amend after review.',
                make_noise_lines(40, 'staging'),
            ].join('\n'),
        },
        'skill-script': {
            tool_name: 'read',
            text: [
                'case "$command" in',
                '  update) run_gh_skill_update "$repo" ;;',
                '  install) run_gh_skill_install "$repo" ;;',
                'esac',
                make_noise_lines(40, 'script'),
            ].join('\n'),
        },
        'code-symbols': {
            tool_name: 'read',
            text: [
                "pi.registerCommand('context-stats', {",
                "  description: 'Show context sidecar byte accounting',",
                '});',
                make_noise_lines(40, 'code'),
            ].join('\n'),
        },
        'chunk-boundary': {
            tool_name: 'mcp__mcp-omnisearch__web_extract',
            text: [
                `alpha-token ${'a'.repeat(5000)}`,
                '',
                `omega-token ${'b'.repeat(5000)}`,
            ].join('\n'),
        },
    };
}
function seed(store) {
    const seeded = new Map();
    for (const [name, source] of Object.entries(source_texts())) {
        const stored = store.store({
            text: source.text,
            tool_name: source.tool_name,
            input_summary: name,
            project_path: PROJECT,
            session_id: SESSION,
            force: true,
        });
        if (!stored)
            throw new Error(`Failed to seed ${name}`);
        seeded.set(name, {
            name,
            source_id: stored.source_id,
            first_chunk_id: stored.first_chunk_id,
        });
    }
    return seeded;
}
function result_bytes(values) {
    return values.reduce((total, value) => total + Buffer.byteLength(value.content, 'utf8'), 0);
}
function push_result(results, result) {
    results.push(result);
}
function run_search_evals(store, seeded, results) {
    for (const scenario of search_scenarios) {
        const source = seeded.get(scenario.source);
        if (!source)
            throw new Error(`Missing source ${scenario.source}`);
        const search = store.search(scenario.query, {
            source_id: source.source_id,
            limit: scenario.limit,
        });
        const passed = search.some((result) => result.content.includes(scenario.expect));
        push_result(results, {
            name: scenario.name,
            category: 'search',
            passed,
            description: scenario.description,
            detail: passed
                ? `matched ${search[0]?.chunk_id ?? 'unknown chunk'}`
                : `no expected match for query ${JSON.stringify(scenario.query)} (${search.length} result(s))`,
            returned_bytes: result_bytes(search),
            result_count: search.length,
        });
    }
}
function run_retrieval_evals(store, seeded, results) {
    for (const scenario of get_scenarios) {
        const source = seeded.get(scenario.source);
        if (!source)
            throw new Error(`Missing source ${scenario.source}`);
        const chunks = store.get(source.source_id, scenario.chunk_id);
        const passed = chunks.some((chunk) => chunk.content.includes(scenario.expect));
        push_result(results, {
            name: scenario.name,
            category: 'retrieval',
            passed,
            description: scenario.description,
            detail: passed
                ? `retrieved ${chunks[0]?.id ?? scenario.chunk_id}`
                : `missing expected content via chunk_id ${scenario.chunk_id}`,
            returned_bytes: result_bytes(chunks),
            result_count: chunks.length,
        });
    }
}
function run_lifecycle_evals(store, seeded, results) {
    const source = seeded.get('staging-help');
    if (!source)
        throw new Error('Missing source staging-help');
    store.purge({ source_id: source.source_id, global: true });
    const chunks = store.get(source.source_id, undefined, { global: true });
    const summary = store.chunk_summary(source.source_id, { global: true });
    push_result(results, {
        name: 'missing-source-guidance',
        category: 'lifecycle',
        passed: chunks.length === 0 && summary !== null,
        description: 'Expired receipt retrieval should distinguish a missing backing source from an empty result and give guidance.',
        detail: summary === null
            ? 'missing source is currently indistinguishable from no data'
            : 'missing source still has retrievable metadata',
        returned_bytes: result_bytes(chunks),
        result_count: chunks.length,
    });
}
function run_capture_evals(results) {
    const large_text = 'x'.repeat(24 * 1024 + 1);
    const many_lines = Array.from({ length: 301 }, (_, index) => String(index)).join('\n');
    const small_text = 'small output';
    const passed = should_index_text(large_text) &&
        should_index_text(many_lines) &&
        !should_index_text(small_text);
    push_result(results, {
        name: 'capture-thresholds',
        category: 'capture',
        passed,
        description: 'Oversized byte and line-count outputs should be captured while small outputs stay inline.',
        detail: passed
            ? 'byte, line, and small-output thresholds behaved as expected'
            : 'capture threshold mismatch',
        returned_bytes: 0,
        result_count: 3,
    });
}
function store_db(store) {
    return Reflect.get(store, 'db');
}
function run_retention_evals(store, results) {
    const old = store.store({
        text: `old-retention-token\n${'o '.repeat(500)}`,
        tool_name: 'bash',
        force: true,
    });
    const fresh = store.store({
        text: `fresh-retention-token\n${'f '.repeat(500)}`,
        tool_name: 'bash',
        force: true,
    });
    if (!old || !fresh)
        throw new Error('Failed to seed retention eval');
    store_db(store)
        .prepare('UPDATE context_sources SET created_at = ? WHERE id = ?')
        .run(Date.now() - 10 * 24 * 60 * 60 * 1000, old.source_id);
    const cleanup = store.cleanup({
        retention_days: 7,
        purge_on_shutdown: false,
        max_mb: null,
        max_bytes: null,
    });
    const old_results = store.search('old-retention-token', {
        source_id: old.source_id,
        global: true,
    });
    const fresh_results = store.search('fresh-retention-token', {
        source_id: fresh.source_id,
        global: true,
    });
    const passed = cleanup.age_deleted >= 1 &&
        old_results.length === 0 &&
        fresh_results.length > 0;
    push_result(results, {
        name: 'age-retention-cleanup',
        category: 'retention',
        passed,
        description: 'Age cleanup should delete expired sources without deleting fresh sources.',
        detail: `age_deleted=${cleanup.age_deleted}; old=${old_results.length}; fresh=${fresh_results.length}`,
        returned_bytes: result_bytes([...old_results, ...fresh_results]),
        result_count: old_results.length + fresh_results.length,
    });
}
function run_cost_evals(store, seeded, results) {
    const source = seeded.get('chunk-boundary');
    if (!source)
        throw new Error('Missing source chunk-boundary');
    const search = store.search('alpha-token omega-token', {
        source_id: source.source_id,
        limit: 25,
    });
    const bytes = result_bytes(search);
    const passed = search.length > 0 && bytes <= 24 * 1024;
    push_result(results, {
        name: 'bounded-relaxed-search-cost',
        category: 'cost',
        passed,
        description: 'Relaxed search should recover useful chunks without dumping excessive context back inline.',
        detail: `results=${search.length}; returned_bytes=${bytes}; budget=24576`,
        returned_bytes: bytes,
        result_count: search.length,
    });
}
function run_dedupe_evals(results) {
    const dir = mkdtempSync(join(tmpdir(), 'pi-context-dedupe-eval-'));
    const store = new ContextStore({
        db_path: join(dir, 'context.db'),
        project_path: PROJECT,
        session_id: `${SESSION}:a`,
    });
    try {
        const text = `shared-dedupe-token\n${'d '.repeat(1000)}`;
        const first = store.store({
            text,
            tool_name: 'read',
            session_id: `${SESSION}:a`,
            project_path: PROJECT,
            force: true,
        });
        const second = store.store({
            text,
            tool_name: 'read',
            session_id: `${SESSION}:b`,
            project_path: PROJECT,
            force: true,
        });
        if (!first || !second)
            throw new Error('Failed to seed dedupe eval');
        const stats = store.stats({ global: true });
        const passed = first.source_id === second.source_id && stats.sources === 1;
        push_result(results, {
            name: 'cross-session-dedupe',
            category: 'dedupe',
            passed,
            description: 'Identical content across sessions should be content-address reused instead of stored twice.',
            detail: `first=${first.source_id}; second=${second.source_id}; sources=${stats.sources}`,
            returned_bytes: 0,
            result_count: stats.sources,
        });
    }
    finally {
        store.close();
        rmSync(dir, { recursive: true, force: true });
    }
}
function build_sections(results) {
    const categories = [
        'search',
        'retrieval',
        'lifecycle',
        'capture',
        'retention',
        'cost',
        'dedupe',
    ];
    return categories
        .map((category) => {
        const section = results.filter((result) => result.category === category);
        const passed = section.filter((result) => result.passed).length;
        const total = section.length;
        return {
            category,
            passed,
            failed: total - passed,
            total,
            score_pct: total === 0
                ? 0
                : Math.round((passed / total) * 1000) / 10,
        };
    })
        .filter((section) => section.total > 0);
}
export function run_context_eval() {
    const dir = mkdtempSync(join(tmpdir(), 'pi-context-eval-'));
    const db_path = join(dir, 'context.db');
    const store = new ContextStore({
        db_path,
        project_path: PROJECT,
        session_id: SESSION,
    });
    try {
        const seeded = seed(store);
        const results = [];
        run_search_evals(store, seeded, results);
        run_retrieval_evals(store, seeded, results);
        run_lifecycle_evals(store, seeded, results);
        run_capture_evals(results);
        run_retention_evals(store, results);
        run_cost_evals(store, seeded, results);
        run_dedupe_evals(results);
        const passed = results.filter((result) => result.passed).length;
        const total = results.length;
        const returned_bytes = results.reduce((sum, result) => sum + result.returned_bytes, 0);
        return {
            version: 1,
            summary: {
                passed,
                failed: total - passed,
                total,
                score_pct: Math.round((passed / total) * 1000) / 10,
                returned_bytes,
            },
            sections: build_sections(results),
            results,
        };
    }
    finally {
        store.close();
        rmSync(dir, { recursive: true, force: true });
    }
}
function format_report(report) {
    return [
        '# pi-context eval',
        '',
        `Score: ${report.summary.passed}/${report.summary.total} (${report.summary.score_pct}%)`,
        `Returned bytes: ${report.summary.returned_bytes}`,
        '',
        'Sections:',
        ...report.sections.map((section) => `- ${section.category}: ${section.passed}/${section.total} (${section.score_pct}%)`),
        '',
        ...report.results.map((result) => [
            `${result.passed ? '✅' : '❌'} ${result.name} (${result.category})`,
            `   ${result.description}`,
            `   ${result.detail}; results=${result.result_count}; bytes=${result.returned_bytes}`,
        ].join('\n')),
    ].join('\n');
}
export async function run_context_eval_cli(args = process.argv.slice(2)) {
    const json = args.includes('--json');
    const report = run_context_eval();
    process.stdout.write(json
        ? `${JSON.stringify(report, null, 2)}\n`
        : `${format_report(report)}\n`);
    if (report.summary.failed > 0)
        process.exitCode = 1;
}
