// @spences10/pi-tui-modal/dist/modal/bodies.js
import { fuzzyFilter, getKeybindings, Input, Key, matchesKey, truncateToWidth, visibleWidth, wrapTextWithAnsi, } from '@earendil-works/pi-tui';
import { normalize_text } from './layout.js';
function expand_tabs(text) {
    return text.replace(/\t/g, '   ');
}
export class TextModalBody {
    text;
    max_visible_lines;
    theme;
    on_cancel;
    offset = 0;
    wrapped_lines = [];
    constructor(text, max_visible_lines, theme, on_cancel) {
        this.text = text;
        this.max_visible_lines = max_visible_lines;
        this.theme = theme;
        this.on_cancel = on_cancel;
    }
    set_max_visible_lines(max_visible_lines) {
        this.max_visible_lines = Math.max(1, max_visible_lines);
    }
    render(width) {
        this.wrapped_lines = normalize_text(this.text).flatMap((block) => expand_tabs(block)
            .split('\n')
            .flatMap((line) => line.length === 0 ? [''] : wrapTextWithAnsi(line, width)));
        const max_offset = Math.max(0, this.wrapped_lines.length - this.max_visible_lines);
        this.offset = Math.max(0, Math.min(this.offset, max_offset));
        const end = Math.min(this.offset + this.max_visible_lines, this.wrapped_lines.length);
        const visible = this.wrapped_lines
            .slice(this.offset, end)
            .map((line) => truncateToWidth(line, width));
        if (this.wrapped_lines.length > this.max_visible_lines) {
            visible.push(this.theme.fg('dim', truncateToWidth(`(${this.offset + 1}-${end}/${this.wrapped_lines.length})`, width)));
        }
        return visible;
    }
    invalidate() {
        // No cached rendering.
    }
    handleInput(data) {
        const keybindings = getKeybindings();
        const max_offset = Math.max(0, this.wrapped_lines.length - this.max_visible_lines);
        if (keybindings.matches(data, 'tui.select.up') || data === 'k') {
            this.offset = Math.max(0, this.offset - 1);
        }
        else if (keybindings.matches(data, 'tui.select.down') ||
            data === 'j') {
            this.offset = Math.min(max_offset, this.offset + 1);
        }
        else if (matchesKey(data, Key.home)) {
            this.offset = 0;
        }
        else if (matchesKey(data, Key.end)) {
            this.offset = max_offset;
        }
        else if (keybindings.matches(data, 'tui.select.cancel') ||
            data === 'q') {
            this.on_cancel();
        }
    }
}
export class InputModalBody {
    options;
    theme;
    on_submit;
    on_cancel;
    input = new Input();
    _focused = false;
    get focused() {
        return this._focused;
    }
    set focused(value) {
        this._focused = value;
        this.input.focused = value;
    }
    constructor(options, theme, on_submit, on_cancel) {
        this.options = options;
        this.theme = theme;
        this.on_submit = on_submit;
        this.on_cancel = on_cancel;
        this.input.setValue(options.initial_value ?? '');
        this.input.onSubmit = (value) => {
            const next_value = options.trim === false ? value : value.trim();
            if (!options.allow_empty && next_value.length === 0)
                return;
            this.on_submit(next_value);
        };
        this.input.onEscape = this.on_cancel;
    }
    render(width) {
        const lines = [];
        if (this.options.label) {
            for (const line of wrapTextWithAnsi(this.options.label, width)) {
                lines.push(this.theme.fg('muted', line));
            }
            lines.push('');
        }
        lines.push(...this.input.render(width));
        lines.push('');
        lines.push(this.theme.fg('dim', this.options.allow_empty
            ? 'Enter submits • Esc cancels'
            : 'Enter submits non-empty value • Esc cancels'));
        return lines.map((line) => truncateToWidth(line, width));
    }
    invalidate() {
        this.input.invalidate();
    }
    handleInput(data) {
        this.input.handleInput(data);
    }
}
export class DetailedSettingsList {
    items;
    max_visible;
    theme;
    on_change;
    on_cancel;
    options;
    filtered_items;
    selected_index = 0;
    search_input;
    constructor(items, max_visible, theme, on_change, on_cancel, options) {
        this.items = items;
        this.max_visible = max_visible;
        this.theme = theme;
        this.on_change = on_change;
        this.on_cancel = on_cancel;
        this.options = options;
        this.filtered_items = items;
        if (options.enable_search)
            this.search_input = new Input();
    }
    get_selected_item() {
        return this.get_display_items()[this.selected_index];
    }
    set_max_visible(max_visible) {
        this.max_visible = Math.max(1, max_visible);
    }
    render(width) {
        const lines = [];
        if (this.search_input) {
            lines.push(...this.search_input.render(width));
            lines.push('');
        }
        const display_items = this.get_display_items();
        if (display_items.length === 0) {
            lines.push(truncateToWidth(this.theme.hint(this.items.length === 0
                ? '  No settings available'
                : '  No matching settings'), width));
            this.add_hint_line(lines, width);
            return lines;
        }
        const start_index = Math.max(0, Math.min(this.selected_index - Math.floor(this.max_visible / 2), display_items.length - this.max_visible));
        const end_index = Math.min(start_index + this.max_visible, display_items.length);
        const max_label_width = Math.min(30, Math.max(...this.items.map((item) => visibleWidth(item.label))));
        const max_value_width = Math.min(14, Math.max(...this.items.map((item) => visibleWidth(item.currentValue))));
        for (let index = start_index; index < end_index; index++) {
            const item = display_items[index];
            if (!item)
                continue;
            lines.push(this.render_item(item, index === this.selected_index, width, max_label_width, max_value_width));
        }
        if (start_index > 0 || end_index < display_items.length) {
            lines.push(this.theme.hint(truncateToWidth(`  (${this.selected_index + 1}/${display_items.length})`, width - 2, '')));
        }
        const selected_item = this.get_selected_item();
        if (selected_item?.description) {
            lines.push('');
            for (const line of wrapTextWithAnsi(selected_item.description, width - 4)) {
                lines.push(this.theme.description(`  ${line}`));
            }
        }
        this.add_hint_line(lines, width);
        return lines;
    }
    invalidate() {
        // No cached rendering.
    }
    handleInput(data) {
        const keybindings = getKeybindings();
        const display_items = this.get_display_items();
        if (keybindings.matches(data, 'tui.select.up')) {
            if (display_items.length === 0)
                return;
            this.selected_index =
                this.selected_index === 0
                    ? display_items.length - 1
                    : this.selected_index - 1;
        }
        else if (keybindings.matches(data, 'tui.select.down')) {
            if (display_items.length === 0)
                return;
            this.selected_index =
                this.selected_index === display_items.length - 1
                    ? 0
                    : this.selected_index + 1;
        }
        else if (keybindings.matches(data, 'tui.select.confirm') ||
            data === ' ') {
            this.activate_item();
        }
        else if (keybindings.matches(data, 'tui.select.cancel')) {
            this.on_cancel();
        }
        else if (this.search_input) {
            const sanitized = data.replace(/ /g, '');
            if (!sanitized)
                return;
            this.search_input.handleInput(sanitized);
            this.apply_filter(this.search_input.getValue());
        }
    }
    get_display_items() {
        return this.search_input ? this.filtered_items : this.items;
    }
    render_item(item, selected, width, max_label_width, max_value_width) {
        const prefix = selected ? this.theme.cursor : '  ';
        const label = truncateToWidth(item.label, max_label_width, '…');
        const padded_label = label +
            ' '.repeat(Math.max(0, max_label_width - visibleWidth(label)));
        const value = truncateToWidth(item.currentValue, max_value_width, '');
        const padded_value = value +
            ' '.repeat(Math.max(0, max_value_width - visibleWidth(value)));
        const detail = this.options.detail?.(item) ?? '';
        const line = [
            prefix,
            this.theme.label(padded_label, selected),
            '  ',
            this.theme.value(padded_value, selected),
            detail ? `  ${this.theme.description(detail)}` : '',
        ].join('');
        return truncateToWidth(line, width);
    }
    activate_item() {
        const item = this.get_selected_item();
        if (!item?.values || item.values.length === 0)
            return;
        const current_index = item.values.indexOf(item.currentValue);
        const next_index = (current_index + 1) % item.values.length;
        const new_value = item.values[next_index];
        item.currentValue = new_value;
        this.on_change(item.id, new_value);
    }
    apply_filter(query) {
        this.filtered_items = fuzzyFilter(this.items, query, (item) => [
            item.label,
            item.currentValue,
            item.description,
            this.options.detail?.(item),
        ]
            .filter(Boolean)
            .join(' '));
        this.selected_index = 0;
    }
    add_hint_line(lines, width) {
        lines.push('');
        lines.push(truncateToWidth(this.theme.hint(this.search_input
            ? '  Type to search · Enter/Space to change · Esc to cancel'
            : '  Enter/Space to change · Esc to cancel'), width));
    }
}
