// @spences10/pi-tui-modal/dist/modal/layout.js
import { wrapTextWithAnsi, } from '@earendil-works/pi-tui';
export const default_overlay_options = {
    width: '80%',
    minWidth: 60,
    maxHeight: '80%',
};
export const default_modal_style = {
    border: 'rounded',
    border_color: 'accent',
};
export const border_characters = {
    rounded: {
        top_left: '╭',
        top: '─',
        top_right: '╮',
        left: '│',
        right: '│',
        bottom_left: '╰',
        bottom: '─',
        bottom_right: '╯',
    },
    square: {
        top_left: '┌',
        top: '─',
        top_right: '┐',
        left: '│',
        right: '│',
        bottom_left: '└',
        bottom: '─',
        bottom_right: '┘',
    },
};
export function normalize_text(value) {
    if (!value)
        return [];
    const resolved = typeof value === 'function' ? value() : value;
    return Array.isArray(resolved) ? resolved : [resolved];
}
export function parse_size_value(value, total) {
    if (value === undefined)
        return undefined;
    if (typeof value === 'number')
        return value;
    const match = value.match(/^(\d+(?:\.\d+)?)%$/);
    return match
        ? Math.floor((Number(match[1]) / 100) * total)
        : undefined;
}
export function get_vertical_margin(margin) {
    if (typeof margin === 'number')
        return Math.max(0, margin) * 2;
    return (Math.max(0, margin?.top ?? 0) + Math.max(0, margin?.bottom ?? 0));
}
export function get_terminal_rows(tui) {
    const rows = tui
        .terminal?.rows;
    return rows ?? process.stdout.rows ?? 24;
}
export function get_border_line_count(style) {
    return style?.border === 'none' ? 0 : 2;
}
export function count_text_lines(value, width) {
    return normalize_text(value).reduce((count, text) => {
        if (text.trim() === '')
            return count;
        return (count +
            wrapTextWithAnsi(text.replace(/\t/g, '   '), width).length);
    }, 0);
}
export function get_modal_body_line_budget(tui, options, body_width = 80) {
    const terminal_rows = get_terminal_rows(tui);
    const overlay_options = {
        ...default_overlay_options,
        ...options.overlay_options,
    };
    const available_height = Math.max(1, terminal_rows - get_vertical_margin(overlay_options.margin));
    const max_height = Math.max(1, Math.min(parse_size_value(overlay_options.maxHeight, terminal_rows) ??
        available_height, available_height));
    const fixed_lines = get_border_line_count(options.style) +
        2 +
        count_text_lines(options.title, body_width) +
        count_text_lines(options.subtitle, body_width) +
        count_text_lines(options.footer, body_width);
    return Math.max(1, max_height - fixed_lines);
}
export function fit_visible_items(item_count, preferred, body_line_budget) {
    const budget = Math.max(1, body_line_budget);
    const candidate = Math.max(1, Math.min(preferred, item_count));
    if (item_count > candidate) {
        return Math.max(1, Math.min(candidate, budget - 1));
    }
    return Math.max(1, Math.min(candidate, budget));
}
export function set_component_max_visible(component, max_visible) {
    component.maxVisible = max_visible;
}
export function normalize_metadata(value, item) {
    if (!value)
        return [];
    const resolved = typeof value === 'function' ? value(item) : value;
    if (!resolved)
        return [];
    return Array.isArray(resolved) ? resolved : [resolved];
}
export function make_select_theme(theme) {
    return {
        selectedPrefix: (text) => theme.fg('accent', text),
        selectedText: (text) => theme.fg('accent', text),
        description: (text) => theme.fg('muted', text),
        scrollInfo: (text) => theme.fg('dim', text),
        noMatch: (text) => theme.fg('warning', text),
    };
}
export function value_color(value) {
    const normalized = value.trim().toLowerCase();
    if (normalized.startsWith('●') ||
        normalized.startsWith('✓') ||
        normalized.includes('enabled') ||
        normalized.includes('selected') ||
        normalized.includes('imported')) {
        return 'success';
    }
    if (normalized.startsWith('↻') ||
        normalized.includes('sync') ||
        normalized.includes('queued')) {
        return 'warning';
    }
    return 'dim';
}
export function get_selected_setting(list) {
    const internals = list;
    const items = internals.searchEnabled
        ? internals.filteredItems
        : internals.items;
    return items[internals.selectedIndex];
}
export function make_settings_theme(theme) {
    return {
        cursor: theme.fg('accent', '→ '),
        label: (text, selected) => {
            if (text.startsWith('──') && text.endsWith('──')) {
                return theme.fg('dim', theme.bold(text));
            }
            return selected ? theme.fg('accent', text) : text;
        },
        value: (text, selected) => {
            const rendered = theme.fg(value_color(text), text);
            return selected
                ? theme.bold(theme.fg('accent', rendered))
                : rendered;
        },
        description: (text) => theme.fg('muted', text),
        hint: (text) => theme.fg('dim', text),
    };
}
export function is_focusable(value) {
    return Boolean(value && typeof value === 'object' && 'focused' in value);
}
