// @spences10/pi-settings/dist/index.js
import { getAgentDir } from '@earendil-works/pi-coding-agent';
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync, } from 'node:fs';
import { dirname, join } from 'node:path';
export function get_settings_path() {
    return join(getAgentDir(), 'my-pi-settings.json');
}
function default_settings() {
    return {
        version: 1,
        extensions: { enabled: {} },
        trust: {},
        packages: {},
    };
}
export function read_settings() {
    const path = get_settings_path();
    if (!existsSync(path))
        return default_settings();
    return {
        ...default_settings(),
        ...JSON.parse(readFileSync(path, 'utf-8')),
    };
}
export function write_settings(settings) {
    const path = get_settings_path();
    mkdirSync(dirname(path), { recursive: true, mode: 0o700 });
    const tmp = `${path}.tmp-${Date.now()}`;
    writeFileSync(tmp, JSON.stringify({ ...settings, version: 1 }, null, '\t') + '\n', { mode: 0o600 });
    renameSync(tmp, path);
}
export function read_settings_section(key, fallback) {
    const settings = read_settings();
    return (settings[key] === undefined ? fallback : settings[key]);
}
export function write_settings_section(key, value) {
    write_settings({ ...read_settings(), [key]: value });
}
export function read_package_settings(name, fallback) {
    const packages = read_settings().packages ?? {};
    return (packages[name] === undefined ? fallback : packages[name]);
}
export function write_package_settings(name, value) {
    const settings = read_settings();
    write_settings({
        ...settings,
        packages: { ...settings.packages, [name]: value },
    });
}
export function read_trust_settings(name, fallback) {
    const trust = read_settings().trust ?? {};
    return (trust[name] === undefined ? fallback : trust[name]);
}
export function write_trust_settings(name, value) {
    const settings = read_settings();
    write_settings({
        ...settings,
        trust: { ...settings.trust, [name]: value },
    });
}
